File: src/C3.Infrastructure/Updates/SemanticVersion.vb
Namespace Updates

    ''' <summary>
    ''' Represents the SemVer release identity used by C3 update manifests.
    ''' Build metadata is retained for display but never affects precedence.
    ''' </summary>
    Public NotInheritable Class SemanticVersion
        Implements IComparable(Of SemanticVersion)

        Public Const MaximumTextCharacters As Integer = 128

        Private ReadOnly _coreIdentifiers As String()
        Private ReadOnly _prereleaseIdentifiers As String()

        Private Sub New(
                originalText As String,
                releaseLabel As String,
                coreIdentifiers As String(),
                prereleaseIdentifiers As String())

            Me.OriginalText = originalText
            Me.ReleaseLabel = releaseLabel
            _coreIdentifiers = coreIdentifiers
            _prereleaseIdentifiers = prereleaseIdentifiers
        End Sub

        Public ReadOnly Property OriginalText As String

        ''' <summary>
        ''' Gets the canonical release label without optional build metadata.
        ''' </summary>
        Public ReadOnly Property ReleaseLabel As String

        Public ReadOnly Property CoreVersion As String
            Get
                Return String.Join(".", _coreIdentifiers)
            End Get
        End Property

        Public ReadOnly Property HasPrerelease As Boolean
            Get
                Return _prereleaseIdentifiers.Length > 0
            End Get
        End Property

        Public ReadOnly Property PrereleaseLabel As String
            Get
                Return String.Join(".", _prereleaseIdentifiers)
            End Get
        End Property

        Public ReadOnly Property FirstPrereleaseIdentifier As String
            Get
                If _prereleaseIdentifiers.Length = 0 Then
                    Return Nothing
                End If
                Return _prereleaseIdentifiers(0)
            End Get
        End Property

        Public Shared Function TryParse(
                value As String,
                ByRef parsed As SemanticVersion) As Boolean

            parsed = Nothing
            If value Is Nothing OrElse value.Length = 0 OrElse
                    value.Length > MaximumTextCharacters OrElse value <> value.Trim() Then
                Return False
            End If

            Dim plusIndex As Integer = value.IndexOf("+"c)
            If plusIndex >= 0 AndAlso value.IndexOf("+"c, plusIndex + 1) >= 0 Then
                Return False
            End If

            Dim releaseLabel As String = If(plusIndex >= 0, value.Substring(0, plusIndex), value)
            If plusIndex >= 0 Then
                Dim buildLabel As String = value.Substring(plusIndex + 1)
                If Not AreValidIdentifiers(buildLabel, False) Then
                    Return False
                End If
            End If

            Dim dashIndex As Integer = releaseLabel.IndexOf("-"c)
            Dim coreLabel As String = If(
                dashIndex >= 0,
                releaseLabel.Substring(0, dashIndex),
                releaseLabel)
            Dim prereleaseLabel As String = If(
                dashIndex >= 0,
                releaseLabel.Substring(dashIndex + 1),
                String.Empty)

            Dim coreIdentifiers As String() = coreLabel.Split("."c)
            If coreIdentifiers.Length <> 3 Then
                Return False
            End If
            For Each identifier As String In coreIdentifiers
                If Not IsCanonicalNumericIdentifier(identifier) Then
                    Return False
                End If
            Next

            Dim prereleaseIdentifiers As String() = New String() {}
            If dashIndex >= 0 Then
                If Not AreValidIdentifiers(prereleaseLabel, True) Then
                    Return False
                End If
                prereleaseIdentifiers = prereleaseLabel.Split("."c)
            End If

            parsed = New SemanticVersion(
                value,
                releaseLabel,
                coreIdentifiers,
                prereleaseIdentifiers)
            Return True
        End Function

        Public Function CompareTo(other As SemanticVersion) As Integer _
                Implements IComparable(Of SemanticVersion).CompareTo

            If other Is Nothing Then
                Return 1
            End If

            For index As Integer = 0 To _coreIdentifiers.Length - 1
                Dim comparison As Integer = CompareNumericIdentifiers(
                    _coreIdentifiers(index),
                    other._coreIdentifiers(index))
                If comparison <> 0 Then
                    Return comparison
                End If
            Next

            If Not HasPrerelease AndAlso Not other.HasPrerelease Then
                Return 0
            End If
            If Not HasPrerelease Then
                Return 1
            End If
            If Not other.HasPrerelease Then
                Return -1
            End If

            Dim sharedLength As Integer = Math.Min(
                _prereleaseIdentifiers.Length,
                other._prereleaseIdentifiers.Length)
            For index As Integer = 0 To sharedLength - 1
                Dim left As String = _prereleaseIdentifiers(index)
                Dim right As String = other._prereleaseIdentifiers(index)
                Dim leftIsNumeric As Boolean = IsNumericIdentifier(left)
                Dim rightIsNumeric As Boolean = IsNumericIdentifier(right)
                Dim comparison As Integer

                If leftIsNumeric AndAlso rightIsNumeric Then
                    comparison = CompareNumericIdentifiers(left, right)
                ElseIf leftIsNumeric Then
                    comparison = -1
                ElseIf rightIsNumeric Then
                    comparison = 1
                Else
                    comparison = String.CompareOrdinal(left, right)
                End If

                If comparison <> 0 Then
                    Return comparison
                End If
            Next

            Return _prereleaseIdentifiers.Length.CompareTo(
                other._prereleaseIdentifiers.Length)
        End Function

        Private Shared Function AreValidIdentifiers(
                value As String,
                enforceNumericCanonicalForm As Boolean) As Boolean

            If value.Length = 0 Then
                Return False
            End If

            Dim identifiers As String() = value.Split("."c)
            For Each identifier As String In identifiers
                If identifier.Length = 0 Then
                    Return False
                End If
                For Each character As Char In identifier
                    If Not IsAsciiLetterOrDigit(character) AndAlso character <> "-"c Then
                        Return False
                    End If
                Next
                If enforceNumericCanonicalForm AndAlso IsNumericIdentifier(identifier) AndAlso
                        Not IsCanonicalNumericIdentifier(identifier) Then
                    Return False
                End If
            Next
            Return True
        End Function

        Private Shared Function IsCanonicalNumericIdentifier(value As String) As Boolean
            Return IsNumericIdentifier(value) AndAlso
                (value.Length = 1 OrElse value(0) <> "0"c)
        End Function

        Private Shared Function IsNumericIdentifier(value As String) As Boolean
            If value.Length = 0 Then
                Return False
            End If
            For Each character As Char In value
                If character < "0"c OrElse character > "9"c Then
                    Return False
                End If
            Next
            Return True
        End Function

        Private Shared Function IsAsciiLetterOrDigit(value As Char) As Boolean
            Return (value >= "0"c AndAlso value <= "9"c) OrElse
                (value >= "A"c AndAlso value <= "Z"c) OrElse
                (value >= "a"c AndAlso value <= "z"c)
        End Function

        Private Shared Function CompareNumericIdentifiers(
                left As String,
                right As String) As Integer

            Dim lengthComparison As Integer = left.Length.CompareTo(right.Length)
            If lengthComparison <> 0 Then
                Return lengthComparison
            End If
            Return String.CompareOrdinal(left, right)
        End Function

    End Class

End Namespace

File: src/C3.Infrastructure/Updates/SemanticVersion.cs
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Runtime.CompilerServices;

namespace C3.Infrastructure.Updates
{
	public sealed class SemanticVersion : IComparable<SemanticVersion>
	{
		public const int MaximumTextCharacters = 128;

		private readonly string[] _coreIdentifiers;

		private readonly string[] _prereleaseIdentifiers;
		public string OriginalText
		{
			get;
		}

		public string ReleaseLabel
		{
			get;
		}

		public string CoreVersion => string.Join(".", _coreIdentifiers);

		public bool HasPrerelease => _prereleaseIdentifiers.Length > 0;

		public string PrereleaseLabel => string.Join(".", _prereleaseIdentifiers);

		public string FirstPrereleaseIdentifier
		{
			get
			{
				if (_prereleaseIdentifiers.Length == 0)
				{
					return null;
				}
				return _prereleaseIdentifiers[0];
			}
		}

		private SemanticVersion(string originalText, string releaseLabel, string[] coreIdentifiers, string[] prereleaseIdentifiers)
		{
			OriginalText = originalText;
			ReleaseLabel = releaseLabel;
			_coreIdentifiers = coreIdentifiers;
			_prereleaseIdentifiers = prereleaseIdentifiers;
		}

		public static bool TryParse(string value, ref SemanticVersion parsed)
		{
			parsed = null;
			checked
			{
				if (value != null && value.Length != 0 && value.Length <= 128 && Operators.CompareString(value, value.Trim(), false) == 0)
				{
					int num = value.IndexOf('+');
					if (num >= 0 && value.IndexOf('+', num + 1) >= 0)
					{
						return false;
					}
					string text = (num >= 0) ? value.Substring(0, num) : value;
					if (num >= 0 && !AreValidIdentifiers(value.Substring(num + 1), false))
					{
						return false;
					}
					int num2 = text.IndexOf('-');
					string obj = (num2 >= 0) ? text.Substring(0, num2) : text;
					string text2 = (num2 >= 0) ? text.Substring(num2 + 1) : string.Empty;
					string[] array = obj.Split('.');
					if (array.Length != 3)
					{
						return false;
					}
					string[] array2 = array;
					for (int i = 0; i < array2.Length; i++)
					{
						if (!IsCanonicalNumericIdentifier(array2[i]))
						{
							return false;
						}
					}
					string[] prereleaseIdentifiers = new string[0];
					if (num2 >= 0)
					{
						if (!AreValidIdentifiers(text2, true))
						{
							return false;
						}
						prereleaseIdentifiers = text2.Split('.');
					}
					parsed = new SemanticVersion(value, text, array, prereleaseIdentifiers);
					return true;
				}
				return false;
			}
		}

		public int CompareTo(SemanticVersion other)
		{
			if (other == null)
			{
				return 1;
			}
			checked
			{
				int num = _coreIdentifiers.Length - 1;
				for (int i = 0; i <= num; i++)
				{
					int num2 = CompareNumericIdentifiers(_coreIdentifiers[i], other._coreIdentifiers[i]);
					if (num2 != 0)
					{
						return num2;
					}
				}
				if (!HasPrerelease && !other.HasPrerelease)
				{
					return 0;
				}
				if (!HasPrerelease)
				{
					return 1;
				}
				if (!other.HasPrerelease)
				{
					return -1;
				}
				int num3 = Math.Min(_prereleaseIdentifiers.Length, other._prereleaseIdentifiers.Length) - 1;
				for (int j = 0; j <= num3; j++)
				{
					string text = _prereleaseIdentifiers[j];
					string text2 = other._prereleaseIdentifiers[j];
					bool flag = IsNumericIdentifier(text);
					bool flag2 = IsNumericIdentifier(text2);
					int num4 = (!flag || !flag2) ? ((!flag) ? (flag2 ? 1 : string.CompareOrdinal(text, text2)) : (-1)) : CompareNumericIdentifiers(text, text2);
					if (num4 != 0)
					{
						return num4;
					}
				}
				return _prereleaseIdentifiers.Length.CompareTo(other._prereleaseIdentifiers.Length);
			}
		}

		int IComparable<SemanticVersion>.CompareTo(SemanticVersion other)
		{
			//ILSpy generated this explicit interface implementation from .override directive in CompareTo
			return this.CompareTo(other);
		}

		private static bool AreValidIdentifiers(string value, bool enforceNumericCanonicalForm)
		{
			if (value.Length == 0)
			{
				return false;
			}
			string[] array = value.Split('.');
			int num = 0;
			bool result;
			while (true)
			{
				if (num < array.Length)
				{
					string text = array[num];
					if (text.Length == 0)
					{
						result = false;
						break;
					}
					string text2 = text;
					foreach (char c in text2)
					{
						if (!IsAsciiLetterOrDigit(c) && c != '-')
						{
							return false;
						}
					}
					if (enforceNumericCanonicalForm && IsNumericIdentifier(text) && !IsCanonicalNumericIdentifier(text))
					{
						return false;
					}
					num = checked(num + 1);
					continue;
				}
				return true;
			}
			return result;
		}

		private static bool IsCanonicalNumericIdentifier(string value)
		{
			if (IsNumericIdentifier(value))
			{
				if (value.Length != 1)
				{
					return value[0] != '0';
				}
				return true;
			}
			return false;
		}

		private static bool IsNumericIdentifier(string value)
		{
			if (value.Length == 0)
			{
				return false;
			}
			foreach (char c in value)
			{
				if (c < '0' || c > '9')
				{
					return false;
				}
			}
			return true;
		}

		private static bool IsAsciiLetterOrDigit(char value)
		{
			if (value >= '0' && value <= '9')
			{
				goto IL_0024;
			}
			if (value >= 'A' && value <= 'Z')
			{
				goto IL_0024;
			}
			if (value >= 'a')
			{
				return value <= 'z';
			}
			return false;
			IL_0024:
			return true;
		}

		private static int CompareNumericIdentifiers(string left, string right)
		{
			int num = left.Length.CompareTo(right.Length);
			if (num != 0)
			{
				return num;
			}
			return string.CompareOrdinal(left, right);
		}
	}
}
